Mathlib/SetTheory/Ordinal/Nimber.lean

```diff
@@ -0,0 +1,361 @@
+  apply le_antisymm <;>
+  apply add_le_of_forall_ne <;>
+  intro x hx
+  on_goal 1 => rw [Nimber.add_comm x, add_ne_add_right]
+  on_goal 2 => rw [Nimber.add_comm a, add_ne_add_left]
+  on_goal 3 => rw [← Nimber.add_comm a, add_ne_add_right]
+  on_goal 4 => rw [← Nimber.add_comm x, add_ne_add_left]
+  all_goals exact hx.ne
+termination_by (a, b)
+
+@[simp]
+theorem add_eq_zero_iff {a b : Nimber} : a + b = 0 ↔ a = b := by
+  constructor <;>
+  intro hab
+  · obtain h | rfl | h := lt_trichotomy a b
+    · have ha : a + a = 0 := add_eq_zero_iff.2 rfl
+      rwa [← ha, add_right_inj, eq_comm] at hab
+    · rfl
+    · have hb : b + b = 0 := add_eq_zero_iff.2 rfl
+      rwa [← hb, add_left_inj] at hab
+  · rw [← Nimber.le_zero]
+    apply add_le_of_forall_ne <;>
+    simp_rw [ne_eq] <;>
+    intro x hx
+    · rw [add_eq_zero_iff, ← hab]
+      exact hx.ne
+    · rw [add_eq_zero_iff, hab]
+      exact hx.ne'
+termination_by (a, b)
+
+theorem add_ne_zero_iff : a + b ≠ 0 ↔ a ≠ b :=
+  add_eq_zero_iff.not
+
+@[simp]
+theorem add_self (a : Nimber) : a + a = 0 :=
+  add_eq_zero_iff.2 rfl
+
+protected theorem add_assoc (a b c : Nimber) : a + b + c = a + (b + c) := by
+  apply le_antisymm <;>
+  apply add_le_of_forall_ne <;>
+  intro x hx <;>
+  try obtain ⟨y, hy, rfl⟩ | ⟨y, hy, rfl⟩ := exists_of_lt_add hx
+  on_goal 1 => rw [Nimber.add_assoc y, add_ne_add_left]
+  on_goal 2 => rw [Nimber.add_assoc _ y, add_ne_add_right, add_ne_add_left]
+  on_goal 3 => rw [Nimber.add_assoc _ _ x, add_ne_add_right, add_ne_add_right]
+  on_goal 4 => rw [← Nimber.add_assoc x, add_ne_add_left, add_ne_add_left]
+  on_goal 5 => rw [← Nimber.add_assoc _ y, add_ne_add_left, add_ne_add_right]
+  on_goal 6 => rw [← Nimber.add_assoc _ _ y, add_ne_add_right]
+  all_goals apply ne_of_lt; assumption
+termination_by (a, b, c)
+
+protected theorem add_zero (a : Nimber) : a + 0 = a := by
+  apply le_antisymm
+  · apply add_le_of_forall_ne
+    · intro a' ha
+      rw [Nimber.add_zero]
+      exact ha.ne
+    · intro _ h
+      exact (Nimber.not_lt_zero _ h).elim
+  · -- by_contra! doesn't work for whatever reason.
+    by_contra h
+    replace h := lt_of_not_le h
+    have := Nimber.add_zero (a + 0)
+    rw [add_left_inj] at this
+    exact this.not_lt h
+termination_by a
+
+protected theorem zero_add (a : Nimber) : 0 + a = a := by
+  rw [Nimber.add_comm, Nimber.add_zero]
+
+instance : Neg Nimber :=
+  ⟨id⟩
+
+@[simp]
+protected theorem neg_eq (a : Nimber) : -a = a :=
+  rfl
+
+instance : AddCommGroupWithOne Nimber where
+  add_assoc := Nimber.add_assoc
+  add_zero := Nimber.add_zero
+  zero_add := Nimber.zero_add
+  nsmul := nsmulRec
+  zsmul := zsmulRec
+  neg_add_cancel := add_self
+  add_comm := Nimber.add_comm
+
+-- TODO: add CharP 2 instance
+
+theorem add_cancel_right (a b : Nimber) : a + b + b = a := by
+  rw [add_assoc, add_self, add_zero]
+
+theorem add_cancel_left (a b : Nimber) : a + (a + b) = b := by
+  rw [← add_assoc, add_self, zero_add]
```

Comment:
Shouldn't `add_cancel_right` and `add_cancel_left` be `@[simp]`?

Reply:
Sure thing.